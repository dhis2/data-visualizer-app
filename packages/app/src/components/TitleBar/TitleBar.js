import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import i18n from '@dhis2/d2-i18n';
import {
    sGetVisualization,
    DEFAULT_VISUALIZATION,
} from '../../reducers/visualization';
import { sGetCurrent, DEFAULT_CURRENT } from '../../reducers/current';
import { sGetUiInterpretation } from '../../reducers/ui';
import { sGetUiLocale } from '../../reducers/settings';
import formatDate from '../../modules/formatDate';
import styles from './styles/TitleBar.style';

export const TITLE_UNSAVED = i18n.t('Unsaved chart');
export const TITLE_DIRTY = i18n.t('Edited');

export const STATE_EMPTY = 'EMPTY';
export const STATE_SAVED = 'SAVED';
export const STATE_UNSAVED = 'UNSAVED';
export const STATE_DIRTY = 'DIRTY';

const defaultTitleStyle = {
    ...styles.cell,
    ...styles.title,
};

const defaultInterpretationStyle = {
    ...styles.cell,
    ...styles.interpretation,
};

const getTitleState = (visualization, current) => {
    if (current === DEFAULT_CURRENT) {
        return STATE_EMPTY;
    } else if (visualization === DEFAULT_VISUALIZATION) {
        return STATE_UNSAVED;
    } else if (current === visualization) {
        return STATE_SAVED;
    } else {
        return STATE_DIRTY;
    }
};

const getTitleText = (titleState, visualization) => {
    switch (titleState) {
        case STATE_UNSAVED:
            return TITLE_UNSAVED;
        case STATE_SAVED:
        case STATE_DIRTY:
            return visualization.name;
        default:
            return '';
    }
};

const getCustomTitleStyle = titleState => {
    switch (titleState) {
        case STATE_UNSAVED:
            return styles.titleUnsaved;
        default:
            return null;
    }
};

const getSuffix = titleState => {
    switch (titleState) {
        case STATE_DIRTY:
            return (
                <div
                    style={{
                        ...styles.suffix,
                        ...styles.titleDirty,
                    }}
                >{`- ${TITLE_DIRTY}`}</div>
            );
        default:
            return '';
    }
};

export const TitleBar = ({ titleState, titleText, interpretationDate }) => {
    const titleStyle = {
        ...defaultTitleStyle,
        ...getCustomTitleStyle(titleState),
    };

    return titleText ? (
        <div style={styles.titleBar}>
            <div style={titleStyle}>
                {titleText}
                {getSuffix(titleState)}
            </div>
            {interpretationDate && (
                <div style={defaultInterpretationStyle}>
                    {i18n.t(
                        'Viewing interpretation from {{interpretationDate}}',
                        {
                            interpretationDate,
                        }
                    )}
                </div>
            )}
        </div>
    ) : null;
};

TitleBar.propTypes = {
    titleState: PropTypes.string,
    titleText: PropTypes.string,
    interpretationDate: PropTypes.string,
};

const mapStateToProps = state => ({
    visualization: sGetVisualization(state),
    current: sGetCurrent(state),
    interpretation: sGetUiInterpretation(state),
    uiLocale: sGetUiLocale(state),
});

const mergeProps = (stateProps, dispatchProps, ownProps) => {
    const { visualization, current, interpretation, uiLocale } = stateProps;
    const titleState = getTitleState(visualization, current);
    return {
        ...dispatchProps,
        ...ownProps,
        titleState: titleState,
        titleText: getTitleText(titleState, visualization),
        interpretationDate:
            interpretation && interpretation.created
                ? formatDate(interpretation.created, uiLocale)
                : null,
    };
};

export default connect(
    mapStateToProps,
    null,
    mergeProps
)(TitleBar);
