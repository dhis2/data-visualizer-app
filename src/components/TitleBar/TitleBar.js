import i18n from '@dhis2/d2-i18n'
import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'
import {
    STATE_UNSAVED,
    STATE_SAVED,
    STATE_DIRTY,
    getVisualizationState,
} from '../../modules/visualization.js'
import { sGetCurrent } from '../../reducers/current.js'
import { sGetVisualization } from '../../reducers/visualization.js'
import styles from './styles/TitleBar.style.js'

export const getTitleUnsaved = () => i18n.t('Unsaved visualization')
export const getTitleDirty = () => i18n.t('Edited')

const defaultTitleStyle = {
    ...styles.cell,
    ...styles.title,
}

const getTitleText = (titleState, visualization) => {
    switch (titleState) {
        case STATE_UNSAVED:
            return getTitleUnsaved()
        case STATE_SAVED:
        case STATE_DIRTY:
            return visualization.displayName || visualization.name
        default:
            return ''
    }
}

const getCustomTitleStyle = (titleState) => {
    switch (titleState) {
        case STATE_UNSAVED:
            return styles.titleUnsaved
        default:
            return null
    }
}

const getSuffix = (titleState) => {
    switch (titleState) {
        case STATE_DIRTY:
            return (
                <div
                    style={{
                        ...styles.suffix,
                        ...styles.titleDirty,
                    }}
                    data-test="titlebar-dirty"
                >{`- ${getTitleDirty()}`}</div>
            )
        default:
            return ''
    }
}

export const UnconnectedTitleBar = ({ titleState, titleText }) => {
    const titleStyle = {
        ...defaultTitleStyle,
        ...getCustomTitleStyle(titleState),
    }

    return titleText ? (
        <div data-test="titlebar" style={styles.titleBar}>
            <div style={titleStyle}>
                {titleText}
                {getSuffix(titleState)}
            </div>
        </div>
    ) : null
}

UnconnectedTitleBar.propTypes = {
    titleState: PropTypes.string,
    titleText: PropTypes.string,
}

const mapStateToProps = (state) => ({
    visualization: sGetVisualization(state),
    current: sGetCurrent(state),
})

const mergeProps = (stateProps, dispatchProps, ownProps) => {
    const { visualization, current } = stateProps
    const titleState = getVisualizationState(visualization, current)
    return {
        ...dispatchProps,
        ...ownProps,
        titleState: titleState,
        titleText: getTitleText(titleState, visualization),
    }
}

export const TitleBar = connect(
    mapStateToProps,
    null,
    mergeProps
)(UnconnectedTitleBar)
