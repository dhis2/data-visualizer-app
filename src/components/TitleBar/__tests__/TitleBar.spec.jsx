import { render, screen } from '@testing-library/react'
import React from 'react'
import {
    STATE_EMPTY,
    STATE_UNSAVED,
    STATE_SAVED,
    STATE_DIRTY,
} from '../../../modules/visualization.js'
import { UnconnectedTitleBar as TitleBar } from '../TitleBar.jsx'

test('TitleBar does not render anything', () => {
    const props = {
        titleState: STATE_EMPTY,
        titleText: null,
    }
    const { container } = render(<TitleBar {...props} />)

    expect(container).toBeEmptyDOMElement()
})

test('TitleBar renders "unsaved" state', () => {
    const props = {
        titleState: STATE_UNSAVED,
        titleText: 'Unsaved visualization',
    }
    render(<TitleBar {...props} />)
    const titleDiv = screen.getByText('Unsaved visualization')
    expect(titleDiv).toBeInTheDocument()
})

test('TitleBar renders "saved" state', () => {
    const props = {
        titleState: STATE_SAVED,
        titleText: 'Yall',
    }

    render(<TitleBar {...props} />)
    const titleDiv = screen.getByText('Yall')
    expect(titleDiv).toBeInTheDocument()
})

test('TitleBar renders "dirty" state', () => {
    const props = {
        titleState: STATE_DIRTY,
        titleText: 'Yall',
    }
    const { container } = render(<TitleBar {...props} />)
    expect(container).toMatchSnapshot()
})
