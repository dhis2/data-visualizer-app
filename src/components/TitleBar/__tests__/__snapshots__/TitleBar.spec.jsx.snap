// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TitleBar renders "dirty" state 1`] = `
<div>
  <div
    data-test="titlebar"
    style="display: flex; justify-content: center;"
  >
    <div
      style="display: flex; align-items: center; background: rgb(255, 255, 255); padding: 6px; border-radius: 5px; margin: 4px; font-size: 14px;"
    >
      Yall
      <div
        data-test="titlebar-dirty"
        style="padding-inline-start: 4px; color: rgb(74, 87, 104);"
      >
        - Edited
      </div>
    </div>
  </div>
</div>
`;
